### PR TITLE
Merge 'Connections' and 'Connection Details' in node details pane

### PR DESCRIPTION
**Pull request by 2opremio**
* Now the Connections table appears at the end of the pane

* Connection details are expandable (still not implemented in the UI but backwards
compatible).

### CODE REVIEW

**Comment by 2opremio** [comment]
Fixes #322

File: render/detailed_node_test.go [hunk 1]
```diff
@@ -75,33 +63,44 @@ func TestMakeDetailedHostNode(t *testing.T) {
 						Key:        "Host name",
 						ValueMajor: "client.hostname.com",
 						ValueMinor: "",
+						Expandable: false,
```
Comment (tomwilkie) [inline render/detailed_node_test.go]: The nil value is false, so this can be omitted

File: render/detailed_node.go [hunk 2]
```diff
@@ -95,49 +110,35 @@ func MakeDetailedNode(r report.Report, n RenderableNode) DetailedNode {
 		}
 	}
 
-	tables := tables{}
+	// Connections table
 	{
-		rows := []Row{}
+		cRows := []Row{}
```
Comment (tomwilkie) [inline render/detailed_node.go]: why the variable rename?  We try and keep things like this out of PRs that add features.

Comment (2opremio) [inline render/detailed_node.go]: it clashes with the `rows` type used below.

File: render/detailed_node.go [hunk 3]
```diff
@@ -74,6 +73,22 @@ func (t tables) Less(i, j int) bool { return t[i].Rank > t[j].Rank }
 // MakeDetailedNode transforms a renderable node to a detailed node. It uses
 // aggregate metadata, plus the set of origin node IDs, to produce tables.
 func MakeDetailedNode(r report.Report, n RenderableNode) DetailedNode {
+	tables := tables{}
```
Comment (tomwilkie) [inline render/detailed_node.go]: All the connections-table-related code is getting big enough to justify its own function now, don't you think?

File: render/detailed_node.go [hunk 4]
```diff
@@ -46,6 +44,7 @@ type Row struct {
 	Key        string `json:"key"`                   // e.g. Ingress
 	ValueMajor string `json:"value_major"`           // e.g. 25
 	ValueMinor string `json:"value_minor,omitempty"` // e.g. KB/s
+	Expandable bool   `json:expandable,omitempty`    // Whether it can be expanded (hidden by default)
```
Comment (tomwilkie) [inline render/detailed_node.go]: ``` `json:"expandable,omitempty"` ```

File: render/detailed_node.go (hunk 2 above)
Comment (tomwilkie) [inline render/detailed_node.go]: which line?  I don't see it.

Comment (2opremio) [inline render/detailed_node.go]: https://github.com/weaveworks/scope/blob/merge-connection-tables/render/detailed_node.go#L134

Comment (tomwilkie) [inline render/detailed_node.go]: Ah I see.  You should move the sort up to where the connections rows are populated.  Also, rename the rows type to rowSort or something; thats a much smaller rename.

File: render/detailed_node.go (hunk 3 above)
Comment (2opremio) [inline render/detailed_node.go]: Sure, my changes only added a couple of lines, but you are right :)

**Comment by tomwilkie** [comment]
@2opremio please wait for an LGTM before merging.  I have more feedback.

File: render/detailed_node.go [hunk 5]
```diff
@@ -151,6 +103,56 @@ func MakeDetailedNode(r report.Report, n RenderableNode) DetailedNode {
...
+		case rate > 1024:
+			return fmt.Sprintf("%.1f", rate/1024), "KBps"
+		default:
+			return fmt.Sprintf("%.0f", rate), "Bps"
+		}
+	}
+
+	rows := []Row{}
+	if n.EdgeMetadata.MaxConnCountTCP != nil {
+		rows = append(rows, Row{"TCP connections", strconv.FormatUint(*n.EdgeMetadata.MaxConnCountTCP, 10), "", false})
```
Comment (tomwilkie) [inline render/detailed_node.go]: can omit the false.

File: render/detailed_node.go [hunk 6]
```diff
@@ -151,6 +103,56 @@ func MakeDetailedNode(r report.Report, n RenderableNode) DetailedNode {
...
+	}
+	if rate, ok := rate(n.EdgeMetadata.EgressPacketCount); ok {
```
Comment (tomwilkie) [inline render/detailed_node.go]: can omit the false.

File: render/detailed_node.go [hunk 7]
```diff
@@ -151,6 +103,56 @@ func MakeDetailedNode(r report.Report, n RenderableNode) DetailedNode {
...
+		rows = append(rows, Row{"Egress packet rate", fmt.Sprintf("%.0f", rate), "packets/sec", false})
+	}
+	if rate, ok := rate(n.EdgeMetadata.IngressPacketCount); ok {
+		rows = append(rows, Row{"Ingress packet rate", fmt.Sprintf("%.0f", rate), "packets/sec", false})
```
Comment (tomwilkie) [inline render/detailed_node.go]: can omit the false.

File: render/detailed_node.go [hunk 8]
```diff
@@ -151,6 +103,56 @@ func MakeDetailedNode(r report.Report, n RenderableNode) DetailedNode {
...
+	}
+	if rate, ok := rate(n.EdgeMetadata.EgressByteCount); ok {
+		s, unit := shortenByteRate(rate)
+		rows = append(rows, Row{"Egress byte rate", s, unit, false})
```
Comment (tomwilkie) [inline render/detailed_node.go]: can omit the false.

File: render/detailed_node.go [hunk 9]
```diff
@@ -151,6 +103,56 @@ func MakeDetailedNode(r report.Report, n RenderableNode) DetailedNode {
...
+	}
+	if rate, ok := rate(n.EdgeMetadata.IngressByteCount); ok {
+		s, unit := shortenByteRate(rate)
+		rows = append(rows, Row{"Ingress byte rate", s, unit, false})
```
Comment (tomwilkie) [inline render/detailed_node.go]: can omit the false.

File: render/detailed_node_test.go [hunk 10]
```diff
@@ -19,8 +19,8 @@ func TestOriginTable(t *testing.T) {
 			Numeric: false,
 			Rank:    2,
 			Rows: []render.Row{
-				{"Name", "apache", ""},
-				{"PID", test.ServerPID, ""},
+				{"Name", "apache", "", false},
+				{"PID", test.ServerPID, "", false},
```
Comment (tomwilkie) [inline render/detailed_node_test.go]: can omit the falses.

File: render/detailed_node_test.go [hunk 11]
```diff
@@ -28,9 +28,9 @@ func TestOriginTable(t *testing.T) {
 			Numeric: false,
 			Rank:    1,
 			Rows: []render.Row{
-				{"Host name", test.ServerHostName, ""},
-				{"Load", "0.01 0.01 0.01", ""},
-				{"Operating system", "Linux", ""},
+				{"Host name", test.ServerHostName, "", false},
+				{"Load", "0.01 0.01 0.01", "", false},
+				{"Operating system", "Linux", "", false},
```
Comment (tomwilkie) [inline render/detailed_node_test.go]: can omit the falses.

File: render/detailed_node_test.go [hunk 12]
```diff
@@ -122,22 +117,13 @@ func TestMakeDetailedContainerNode(t *testing.T) {
...
 				Title:   "Origin Container",
 				Numeric: false,
 				Rank:    3,
 				Rows: []render.Row{
-					{"ID", test.ServerContainerID, ""},
-					{"Name", "server", ""},
-					{"Image ID", test.ServerContainerImageID, ""},
+					{"ID", test.ServerContainerID, "", false},
+					{"Name", "server", "", false},
+					{"Image ID", test.ServerContainerImageID, "", false},
```
Comment (tomwilkie) [inline render/detailed_node_test.go]: can omit the falses.

File: render/detailed_node_test.go [hunk 13]
```diff
@@ -154,45 +140,54 @@ func TestMakeDetailedContainerNode(t *testing.T) {
 				Numeric: false,
 				Rank:    1,
 				Rows: []render.Row{
-					{"Host name", test.ServerHostName, ""},
-					{"Load", "0.01 0.01 0.01", ""},
-					{"Operating system", "Linux", ""},
+					{"Host name", test.ServerHostName, "", false},
+					{"Load", "0.01 0.01 0.01", "", false},
+					{"Operating system", "Linux", "", false},
```
Comment (tomwilkie) [inline render/detailed_node_test.go]: can omit the false.

File: render/detailed_node.go (hunk 5 above)
Comment (2opremio) [inline render/detailed_node.go]: No I can't:

`../src/github.com/weaveworks/scope/render/detailed_node.go:130: too few values in struct initializer`

File: render/detailed_node.go (hunk 6 above)
Comment (2opremio) [inline render/detailed_node.go]: ?

File: render/detailed_node.go (hunk 8 above)
Comment (2opremio) [inline render/detailed_node.go]: No, I can't `../src/github.com/weaveworks/scope/render/detailed_node.go:136: too few values in struct initializer`

File: render/detailed_node.go (hunk 9 above)
Comment (2opremio) [inline render/detailed_node.go]: Same here

File: render/detailed_node.go (hunk 7 above)
Comment (2opremio) [inline render/detailed_node.go]: I can't `../src/github.com/weaveworks/scope/render/detailed_node.go:136: too few values in struct initializer`

File: render/detailed_node_test.go (hunk 10 above)
Comment (2opremio) [inline render/detailed_node_test.go]: Same here

File: render/detailed_node_test.go (hunk 11 above)
Comment (2opremio) [inline render/detailed_node_test.go]: Same here :)

File: render/detailed_node_test.go (hunk 12 above)
Comment (2opremio) [inline render/detailed_node_test.go]: Idem

File: render/detailed_node_test.go (hunk 13 above)
Comment (2opremio) [inline render/detailed_node_test.go]: Idem :)

File: render/detailed_node.go (hunk 5 above)
Comment (tomwilkie) [inline render/detailed_node.go]: We should probably use struct labels here then, even through thats a bigger change.  unlabelled struct initialisers are bad.

**Pull request merged**